feat: unstable HttpClient Config

Fishrock123 opened the pull request:
This adds an `unstable-config` feature, with a new `Config` struct, which can be used to configure any `HttpClient` which implements support for it.

Currently it supports two features - the most important and most generally supported:
- `timeout` (`Duration`)
- `tcp_no_delay` (`bool`)

Implementations are provided for async-h1, isahc, and hyper (partial, no `no_delay` support due to the tls connector).

No serious attempt has been made to add this to the wasm client at this point, since I don't understand well how to even build the wasm client or if it even works anymore with the state of rust wasm web build tools.

This represents the development of the older draft PR: https://github.com/http-rs/http-client/pull/84
Should address https://github.com/http-rs/http-client/issues/61

----

Possible next targets:
- [x] `http_keep_alive` on/off (Requires a lot of code shuffling in the h1 client)
- [ ] socket `ttl` (Does not appear to be accessible from hyper)
- [x] `max_connections` - a slightly haphazard api by the h1 client
    - Isahc has two options for this: https://docs.rs/isahc/1.3.1/isahc/struct.HttpClientBuilder.html#method.max_connections
    - Currently ours is per-host and probably ought to be renamed, and a client-global one added (somehow)
    - Hyper ... doesn't have comparable options? Idk: https://docs.rs/hyper/0.14.7/hyper/client/struct.Builder.html

Fishrock123 committed:
feat: unstable HttpClient Config

This adds an `unstable-config` feature, with a new `Config` struct, which can be used to configure any `HttpClient` which implements support for it.

Currently it supports two features - the most important and most generally supported:
- `timeout` (`Duration`)
- `no_delay` (`bool`)

Implementations are provided for async-h1, isahc, and hyper (partial, no `no_delay` support due to the tls connector).

No serious attempt has been made to add this to the wasm client at this point, since I don't understand well how to even build the wasm client or if it even works anymore with the state of rust wasm web build tools.

Fishrock123 committed:
feat: add `http_keep_alive` to Config

Also renamed `no_delay` to `tcp_no_delay`.

the options for configuring this in Isahc and Hyper are unfortunately not super clear.

Fishrock123 commented:
Updated with an `http_keep_alive` option.

jbr commented:
I think I've mentioned this before, but I'd be curious how tls configuration fits into this — would that be a later enhancement?

Fishrock123 commented:
~Probably - as it stands a third of our configurations are completely unable to use TLS configuration of any kind, because they use dependencies that offer no options, namely:~
- ~hyper-tls~
- ~async-tls (the rustls adaptor, but rustls itself is very configurable)~

We could support PEM or DER encoded X.509 certificates for isahc and async-native-tls, and maybe a couple extra options:
- isahc: can set ca certs to trust, ssl ciphers, client cert, or be told to trust invalid things
- async-native-tls: can _add_ ca certs to trust, can be told to trust invalid things, ssl ciphers controlled by tls version flags

EDIT: Turns out I was very wrong and these things do indeed have configuration for this but it's a little hidden.

Fishrock123 committed:
feat: add `tls_config` to Config

Only supports the h1 client, but has two different options, one each for native-tls and rustls.

Fishrock123 committed:
refactor: share the H1Client Config via an Arc

Fishrock123 commented:
Ok updated with tls options for the async-h1 client. Different configuration is accepted depending if `native-tls` or `rustls` is selected.

This is not implemented for the isahc or hyper clients as those can be set directly and then wrapped, and the complexity of doing that integration fully would be too substantial for this PR.

Fishrock123 commented:
I'd rather avoid `max_connections` for this unstable pass, it should really be re-thought for `http-client` 7.0...

Fishrock123 commented:
merging this as an unstable feature and releasing it.

Fishrock123 commented:
Released into Surf: https://github.com/http-rs/surf/releases/tag/v2.3.0
